### PR TITLE
perf(input): prevent multiple validations on compilation

### LINKED ISSUE #14691: $asyncValidators run multiple times per element on load (given ng-repeat || ng-(min/max)length))
**_Note_: for support questions, please use one of these channels: https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#question. This repository's issues are reserved for feature requests and bug reports.**

**Do you want to request a _feature_ or report a _bug_?**

_Bug_

**What is the current behavior?**

If you have a model defined already and it's put into an input text element with a custom validator defined and this input field is in an ng-repeat, then, on the page load, the $ayncValidator will be run once, and then run in quick succession run again for every validator directive such as "minlength, maxlength, ng-maxlength, ng-minlength, required".

**If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem via https://plnkr.co or similar (template: http://plnkr.co/edit/tpl:yBpEi4).**
1. define a model in the controller
2. define a custom directive that has an $asyncValidator linked (put in a console.log before returning the promise)
3. create a div with ng-repeat (if this step is omitted, then the problem is only exhibited with ng-maxlength and ng-minlength)
4. Inside of the ng-repeat clause create an input type="text" element, and load it up with maxlength, minlength, and required
5. take note of how many times the console.log() that was set up in $asyncValidator is tripped when you hit "Run". 
6. when you go and interact with the element, then it behaves normally and as expected.

http://jsfiddle.net/yofk7u0L/5/

**What is the expected behavior?**

$asyncValidators only fire once upon directive initialization with a non-empty model, regardless of how many other attribute directives that are validators exist in the element, also regardless of whether they're in ng-repeat or not.

**What is the motivation / use case for changing the behavior?**

In the application I'm currently working on, a user can have up to 10 of these $asyncValidator input fields loaded up at the same time. With minlength, maxlength, and required enabled on each, this makes 30 EXTRA requests to my API, on top of the 10 that were welcome. If I use ng-minlength and ng-maxlength, it could bump it up an extra twenty (two per input element)

**Which versions of Angular, and which browser / OS are affected by this issue? Did this work in previous versions of Angular? Please also test with the latest stable and snapshot (https://code.angularjs.org/snapshot/) versions.**

Firefox and Chrome on Angular 1.5.3, 1.5.5, 1.5.6 and v1.5.7-build.4838+sha.cfc8b41 exhibit this behavior.
I'm seeing this happen on both Windows 7 (1.5.3 & 1.5.5) and Linux Fedora 23 KDE (All the above)

**Other information (e.g. stacktraces, related issues, suggestions how to fix)**
Not sure how to fix this, but I guess:
- Order directives are placed have no bearing
- Synchronous $validators within the same directive as the $asyncValidators behave normally.
- The entire list of $asyncValidators seems to be looped through each time
- I have not tried adding custom directives to the input element, but I suspect that they might behave similarly and contribute to the amount of times
- If you use a combination of ng-repeat and ng-minlength and ng-maxlength that seems to maximize the total number of requests.
- Have not tested to see if other input types are affected

### PR DESCRIPTION
**Pull request by Narretz**
This commit updates in-built validators with observers to prevent
multiple calls to $validate that could happen after compilation in
certain circumstances:

- when an input is wrapped in a transclude: element directive (e.g. ngRepeat),
the order of execution between ngModel and the input / validation directives changes so that
the initial observer call happens when ngModel has already been initalized,
leading to another call to $validate, which calls *all* defined validators again.
Without ngRepeat, ngModel hasn't been initialized yet, and $validate does not call the validators.

When using validators with scope expressions, the expression value is not available when
ngModel first runs the validators (e.g. ngMinlength="myMinlength"). Only in the first call to
the observer does the value become available, making a call to $validate a necessity.

This commit solves the first problem by storing the validation attribute value so we can compare
the current value and the observed value - which will be the same after compilation.

The second problem is solved by parsing the validation expression once in the link function,
so the value is available when ngModel first validates.

Closes #14691

**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**

**What is the current behavior? (You can also link to an open issue here)**

**What is the new behavior (if this is a feature change)?**

**Does this PR introduce a breaking change?**

**Please check if the PR fulfills these requirements**
- [ ] The commit message follows our [guidelines](https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#commits)
- [ ] Fix/Feature: [Docs](https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#documentation) have been added/updated
- [ ] Fix/Feature: Tests have been added; existing tests pass

**Other information**:

### CODE REVIEW

File: src/ng/directive/input.js [hunk 1]
```diff
@@ -1540,24 +1540,35 @@ function createDateInputType(type, regexp, parseDate, format) {
...
+        return !isValidDate(value) || isUndefined(parsedMinVal) || parseDate(value) >= parsedMinVal;
       };
       attr.$observe('min', function(val) {
-        minVal = parseObservedDateValue(val);
-        ctrl.$validate();
+        if (val !== minVal) {
+          parsedMinVal = parseObservedDateValue(val);
+          ctrl.$validate();
+        }
+        minVal = val;
```
Comment (jbedard) [inline src/ng/directive/input.js]: These could go within the `if`.

Any reason to do it after the `$validate` call along with the `parsedMinValue` assignment?

File: src/ng/directive/validators.js [hunk 2]
```diff
@@ -62,20 +62,26 @@
...
  * </example>
  */
-var requiredDirective = function() {
+var requiredDirective = function($parse) {
   return {
     restrict: 'A',
     require: '?ngModel',
     link: function(scope, elm, attr, ctrl) {
       if (!ctrl) return;
+      var oldVal = attr.required || $parse(attr.ngRequired)(scope);
```
Comment (jbedard) [inline src/ng/directive/validators.js]: What if `attr.required` is `false`? 🤔

Comment (petebacondarwin) [reply src/ng/directive/validators.js]: I believe that `attr.required === false` can't occur, right? If the attribute exists at all then, being a boolean attribute, its value will be `true`.

**Review by petebacondarwin — APPROVED** [review]
Generally LGTM. A few minor suggestions.

File: src/ng/directive/validators.js [hunk 3]
```diff
@@ -353,21 +379,48 @@ var maxlengthDirective = function() {
...
 };
+
+
+function parsePatternAttr(regex, patternExp, elm) {
+
+  if (isString(regex) && regex.length > 0) {
+    regex = new RegExp('^' + regex + '$');
+  }
+
+  if (regex && !regex.test) {
```
[the remark this answers is not in the document]
Comment (Narretz) [reply src/ng/directive/validators.js]: I can dig in the code to see if there's a reason for this (not my change ;)

File: src/ng/directive/validators.js [hunk 4]
```diff
@@ -162,27 +168,43 @@ var requiredDirective = function() {
...
+      if (attr.ngPattern) {
+        patternExp = attr.ngPattern;
 
-        if (regex && !regex.test) {
-          throw minErr('ngPattern')('noregexp',
-            'Expected {0} to be a RegExp but was {1}. Element: {2}', patternExp,
-            regex, startingTag(elm));
+        // ngPattern might be a scope expressions, or an inlined regex, which is not parsable.
+        // We get value of the attribute here, so we can compare the old and the new value
+        // in the observer to avoid unnessecary validations
```
[the remark this answers is not in the document]
Comment (Narretz) [reply src/ng/directive/validators.js]: This word is the bane of my existence.

File: test/helpers/testabilityPatch.js [hunk 5]
```diff
@@ -312,7 +312,28 @@ window.dump = function() {
...
     module(function($compileProvider) {
+      $compileProvider.directive('validationSpy', function() {
+        return {
+          priority: 1,
+          require: 'ngModel',
+          link: function(scope, element, attrs, ctrl) {
+            var validationName = attrs.validationSpy;
+
+            var originalValidator = ctrl.$validators[validationName];
+            helper.validationCounter[validationName] = 0;
```
[the remark this answers is not in the document]
Comment (petebacondarwin) [reply test/helpers/testabilityPatch.js]: Right but in those tests you are recompiling so the directive will be linked again, causing the object to be reset. But I guess it does no harm to reset it in the test as well.

File: src/ng/directive/input.js [hunk 6]
```diff
@@ -1540,24 +1540,34 @@ function createDateInputType(type, regexp, parseDate, format) {
...
+      var parsedMinVal = parseObservedDateValue(minVal);
+
       ctrl.$validators.min = function(value) {
-        return !isValidDate(value) || isUndefined(minVal) || parseDate(value) >= minVal;
+        return !isValidDate(value) || isUndefined(parsedMinVal) || parseDate(value) >= parsedMinVal;
       };
       attr.$observe('min', function(val) {
-        minVal = parseObservedDateValue(val);
-        ctrl.$validate();
+        if (val !== minVal) {
```
Comment (gkalpak) [inline src/ng/directive/input.js]: So, if `val` is compared against `minVal` and `ctrl.$validate()` is only run inside `$observe()`, what guarantees that it is run at least once (for the initial `minVal`)?

I guess in most cases, `ngModel` itself runs it, but is it always guaranteed to happen after the `min` validator has been set up?

(I am not saying it isn't, just wondering.)

File: src/ng/directive/validators.js [hunk 7]
```diff
@@ -353,21 +387,49 @@ var maxlengthDirective = function() {
...
+
+
+function parsePatternAttr(regex, patternExp, elm) {
+  if (!regex) return undefined;
+
+  if (isString(regex) && regex.length > 0) {
+    regex = new RegExp('^' + regex + '$');
+  }
+
+  if (!(regex instanceof RegExp)) {
```
Comment (gkalpak) [inline src/ng/directive/validators.js]: Checking for `instanceof RegExp` instead of `regep.test` looks like a breaking change (and an unnecessary one afaict :grin:).

File: src/ng/directive/validators.js [hunk 8]
```diff
@@ -162,34 +167,59 @@ var requiredDirective = function() {
...
     restrict: 'A',
     require: '?ngModel',
-    link: function(scope, elm, attr, ctrl) {
-      if (!ctrl) return;
+    compile: function(tElm, tAttr) {
+      var patternExp;
+      var parseFn;
+
+      if (tAttr.ngPattern) {
+        patternExp = tAttr.ngPattern;
```
Comment (gkalpak) [inline src/ng/directive/validators.js]: Previously, `patternExp` would also be set to `attr.pattern` (if `attr.ngPattern` was undefined).
Why is this changed?

File: src/ng/directive/validators.js [hunk 9]
```diff
@@ -162,34 +167,59 @@ var requiredDirective = function() {
...
-      attr.$observe('pattern', function(regex) {
-        if (isString(regex) && regex.length > 0) {
-          regex = new RegExp('^' + regex + '$');
+        // ngPattern might be a scope expressions, or an inlined regex, which is not parsable.
+        // We get value of the attribute here, so we can compare the old and the new value
+        // in the observer to avoid unnecessary validations
+        try {
+          parseFn = $parse(tAttr.ngPattern);
+        } catch (e) {
+          if (/^\[\$parse:lexerr\]/.test(e.message)) {
```
[the remark this answers is not in the document]
Comment (gkalpak) [reply src/ng/directive/validators.js]: Why not re-use the logic from the `ngPattern` aliased directive?

**Review by gkalpak — COMMENTED** [review]
I generally like the idea of not re-validating if the values haven't changed (as long as we can guarantee that no validations are skipped when they shouldn't).

I am not a big fan of manually parsing values though. Couldn't we avoid that and instead rely on the parsing done inside the observers?

File: src/ng/directive/input.js (hunk 6 above)
Comment (Narretz) [reply src/ng/directive/input.js]: For our validators, it's guaranteed that they will run when ngModel is set up. That's because we have a watch on the model value, so everything that is added in a link fn is guaranteed to be set up before the model value is first set to control / validated in the watch action fn.

File: src/ng/directive/validators.js (hunk 8 above)
Comment (Narretz) [reply src/ng/directive/validators.js]: if ngPattern isn't set, there can be no pattern expression, because pattern can only have a string value.

File: src/ng/directive/validators.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (Narretz) [reply src/ng/directive/validators.js]: It wasn't handled before, because we didn't need to compare the old / new values. This is made problematic in ngPattern because it is special cased to allow RegExp (which is normally not allowed). 

You mean we should use `REGEX_STRING_REGEXP` instead?

File: src/ng/directive/validators.js [hunk 10]
```diff
@@ -353,21 +387,49 @@ var maxlengthDirective = function() {
...
+var minlengthDirective = function($parse) {
   return {
     restrict: 'A',
     require: '?ngModel',
     link: function(scope, elm, attr, ctrl) {
       if (!ctrl) return;
 
-      var minlength = 0;
+      var minlength = attr.minlength || $parse(attr.ngMinlength)(scope);
+      var minlengthParsed = toInt(minlength) || 0;
```
[the remark this answers is not in the document]
Comment (Narretz) [reply src/ng/directive/validators.js]: The tests agree :)

File: src/ng/directive/validators.js (hunk 7 above)
Comment (Narretz) [reply src/ng/directive/validators.js]: We only allow RegExp and regex literals, which both are identified correctly by instanceof. Or could there be another reason for having .test instead?

File: test/ng/directive/validatorsSpec.js [hunk 11]
```diff
@@ -626,5 +700,43 @@ describe('validators', function() {
...
+      helper.compileInput(
+         '<div ng-repeat="input in [0]">' +
+           '<input type="text" ng-model="value" required validation-spy="required" />' +
+         '</div>');
+
+      $rootScope.$digest();
+
+      expect(helper.validationCounter.required).toBe(1);
+
+      helper.validationCounter = {};
```
[the remark this answers is not in the document]
Comment (Narretz) [reply test/ng/directive/validatorsSpec.js]: Not really. Should I remove it?

File: src/ng/directive/validators.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (jbedard) [reply src/ng/directive/validators.js]: This still seems iffy to me :/

File: src/ng/directive/input.js [hunk 12]
```diff
@@ -1709,50 +1719,68 @@ function isValidForStep(viewValue, stepBase, step) {
...
-             isValidForStep(viewValue, minVal || 0, stepVal);
+      return ctrl.$isEmpty(viewValue) || isUndefined(parsedStepVal) ||
+        isValidForStep(viewValue, parsedMinVal || 0, parsedStepVal);
     };
 
     attr.$observe('step', function(val) {
-      stepVal = parseNumberAttrVal(val);
       // TODO(matsko): implement validateLater to reduce number of validations
-      ctrl.$validate();
+      if (val !== stepVal) {
```
Comment (gkalpak) [inline src/ng/directive/input.js]: I suspect that we should be also checking for `minVal` here, because the `step` validation depends on that.

File: src/ng/directive/input.js [hunk 13]
```diff
@@ -1827,7 +1861,13 @@ function rangeInputType(scope, element, attr, ctrl, $sniffer, $browser) {
     // attribute value when the input is first rendered, so that the browser can adjust the
     // input value based on the min/max value
     element.attr(htmlAttrName, attr[htmlAttrName]);
-    attr.$observe(htmlAttrName, changeFn);
+    var oldVal = attr[htmlAttrName];
+    attr.$observe(htmlAttrName, function wrappedObserver(val) {
+      if (val !== oldVal) {
```
Comment (gkalpak) [inline src/ng/directive/input.js]: Same as above: I believe `minVal` should be checked along with `stepVal`.

File: src/ng/directive/validators.js (hunk 8 above)
Comment (gkalpak) [reply src/ng/directive/validators.js]: Theoretically, it can have any value :grin:
(E.g. imagine a custom `myPattern` directive, that sets `$attrs.pattern`.)

File: src/ng/directive/validators.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (gkalpak) [reply src/ng/directive/validators.js]: @Narretz, yes, I suggest we re-use that bit of logic:

https://github.com/angular/angular.js/blob/0687a4f70a16315aacd9e74edaaac466208bb86d/src/ng/directive/attrs.js#L392-L395

File: src/ng/directive/validators.js (hunk 7 above)
Comment (gkalpak) [reply src/ng/directive/validators.js]: I mean someone might be taking advantage of the fact that we were only testing for `.test` and provide a custom object like `{test: () => /* Do some fancy stuff */}`.
I know it is not a great idea, but I also know I've done it in the past :grin:

What I am saying is that I don't see any benefit is breaking this (admittedly dubious) usecase at this point (a few months into LTS already).
Is there any benefit in changing that?

File: test/ng/directive/validatorsSpec.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (gkalpak) [reply test/ng/directive/validatorsSpec.js]: Yes, please :grin:

File: src/ng/directive/input.js (hunk 12 above)
Comment (Narretz) [reply src/ng/directive/input.js]: You mean ensure that the minVal has changed? That doesn't work, because we need to run this also if the minVal is the same.
This is a good point though - the changes in this PR only prevent multiple calls to $validate after compilation. If you have a range or number with more than one modifier, i.e. step, max, min and change all values in a single digest, every observer willcall $validate. 
I don't think there's an easy way to prevent this, because the observers are independent of each other. Maybe that's what Matias meant with validateLater - a way to collect the $validate calls of a single "validation group" (For the record, I thought he meant a function that collects all $validate calls on an element).
Long story short, I don't think we can do anything about this without making bigger changes. I can adjust the commit message to reflect that changing multiple observed values will still result in multiple validate calls. (I believe this happens much rarer though, so the PR is still worthwhile).

Comment (gkalpak) [reply src/ng/directive/input.js]: No, I don't me ensure that `minVal` has changed. What I meant is that that validator needs to run if either of `minVal` or `stepVal` has changed. I.e. even if `stepVal` is the same, but `minVal` has changed.

##
I am pretty sure that indeed Matias meant avoid multiple calls to `$validate()` :grin:
We could easily do this by running `$validate()` asynchronously (e.g. adding it in the `$postDigest` queue and only if it is not already scheduled), but that would be a breaking change (which does not make sense at this point).

We could use a similar technique in the validators as this PR uses in observers:
Keep track of the involved values (e.g. module/viewValue, validators params, etc) and do not re-run validation unless the values have changed. But that would complicate things even further, open up new possibilities for memory leaks (if not handled carefully) and would also not help with the stateful async validators (which are usually the most expensive ones). So, again, not worth it atm.

> I believe this happens much rarer though, so the PR is still worthwhile.

Actually, I believe it will happen all the time 😛
Basically, afaict, saves some (half) of the re-runs during initialization, but does nothing to prevent multiple runs as the user interacts with the input (which is where the majority of runs comes from throughout the lifetime of an input).

If we can keep it clean and BC-free (with reasonable confidence :grin:), then it is definitely worth landing. But I wouldn't go to great extends, complicating the internal logic too much etc., just to save a few runs :grin:
